File: guides/java/kora-java-guide-resilient-app/src/main/java/io/koraframework/guide/resilient/service/CircuitBreakerFailurePredicate.java
package io.koraframework.guide.resilient.service;

import io.koraframework.common.annotation.Component;
import io.koraframework.http.server.common.response.HttpServerResponseException;
import io.koraframework.resilient.circuitbreaker.CircuitBreakerPredicate;

@Component
public final class CircuitBreakerFailurePredicate implements CircuitBreakerPredicate {

    @Override
    public String name() {
        return "RecordServerErrorsOnly";
    }

    @Override
    public boolean test(Throwable throwable) {
        if (throwable instanceof HttpServerResponseException exception) {
            return exception.code() >= 500;
        }
        return true;
    }
}

File: guides/kotlin/kora-kotlin-guide-resilient-app/src/main/kotlin/io/koraframework/guide/resilient/service/CircuitBreakerFailurePredicate.kt
package io.koraframework.guide.resilient.service

import io.koraframework.common.annotation.Component
import io.koraframework.http.server.common.response.HttpServerResponseException
import io.koraframework.resilient.circuitbreaker.CircuitBreakerPredicate

@Component
class CircuitBreakerFailurePredicate : CircuitBreakerPredicate {
    override fun name(): String = "RecordServerErrorsOnly"

    override fun test(throwable: Throwable): Boolean {
        return throwable !is HttpServerResponseException || throwable.code() >= 500
    }
}
